src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -355,6 +355,34 @@ SendCommandListToWorkerOutsideTransaction(const char *nodeName, int32 nodePort,
 }
 
 
+/*
+ * SendCommandListToWorkerInSeparateTransaction sends the command list over the
+ * connection in a separate connection. This opens a new transaction on the
+ * connection, thus it's important that no transaction is currently open on the
+ * given connection. This function is mainly useful to avoid opening an closing
+ * connections excessively by allowing reusing a single connection to send
+ * multiple separately committing transactions. The function raises an error if
+ * any of the queries fail.
+ */
+void
+SendCommandListToWorkerInSeparateTransaction(MultiConnection *workerConnection,
+											 List *commandList)
+{
+	MarkRemoteTransactionCritical(workerConnection);
+	RemoteTransactionBegin(workerConnection);
+
+	/* iterate over the commands and execute them in the same connection */
+	const char *commandString = NULL;
+	foreach_ptr(commandString, commandList)
+	{
+		ExecuteCriticalRemoteCommand(workerConnection, commandString);
+	}
+
+	RemoteTransactionCommit(workerConnection);
+	CloseRemoteTransaction(workerConnection);
```

Comment:
would it make sense to do this in RemoteTransactionCommit? (and other relevant functions)

Reply:
Possibly, but I don't have a lot of knowledge on our connection management code. I'd prefer to wait with merging this until Monday, to get some feedback from @onderkalaci.

Reply:
So, now `SendCommandListToWorkerInSeparateTransaction()` is almost the same with `SendCommandListToWorkerOutsideTransaction()`. The primary difference is FORCE_NEW_CONNECTION flag.

The difference is non-trivial for readers to follow. Can't we use `SendCommandListToWorkerInSeparateTransaction()` inside `SendCommandListToWorkerOutsideTransaction`? That could make it easier to follow and little more robust for future changes.

> would it make sense to do this in RemoteTransactionCommit? (and other relevant functions)

I think that is a good idea, we also need `CloseShardPlacementAssociation` and probably make `CloseShardPlacementAssociation/CloseRemoteTransaction` static functions?

Reply:
> The difference is non-trivial for readers to follow. Can't we use SendCommandListToWorkerInSeparateTransaction() inside SendCommandListToWorkerOutsideTransaction? That could make it easier to follow and little more robust for future changes.

Something like this: https://github.com/citusdata/citus/commit/708a5829f5cca9c482d8383ce5e7035fb850cbfc

Reply:
@marcocitus Can you elaborate on the comment "would it make sense to do this in RemoteTransactionCommit? (and other relevant functions)"

@onderkalaci Why do we also need "CloseShardPlacementAssociation" ? Looks like function is used to handles a connection being closed before transaction end. However we don't want to close connection in this case.

Reply:
RemoteTransactionCommit changes the state of the remote transaction, but apparently not to the right value for reuse. That seems like an issue in FinishRemoteTransactionCommit. It also seems like we should not just change the state of the session without actually doing anything (read: call CloseRemoteTransaction directly). Those two things always need to be coupled, so it makes more sense to me to set the right state in FinishRemoteTransactionCommit. 

Looking at FinishRemoteTransactionCommit though, it is a little weird. It sets the state to REMOTE_TRANS_COMMITTED, which we presumably treat differently than REMOTE_TRANS_NOT_STARTED, but it does not set the state at all in case of error. This probably requires a separate look (cc @onderkalaci )

Maybe easist to change from CloseRemoteTransaction (which is kind of a low-level internal function) to ResetConnection.

Reply:
> Maybe easist for now to change from CloseRemoteTransaction (which is kind of a low-level internal function) to ResetConnection in this line.

makes sense